File: backend/methods/rk4.py
from .utils import evaluate_function, evaluate_system

def runge_kutta_4(func_str, x0, y0, h, steps_count):
    """
    Runge-Kutta 4th Order Method
    """
    steps = []
    x = x0
    y = y0
    
    for i in range(steps_count):
        try:
            k1 = h * evaluate_system([func_str], {'x': x, 'y': y})[0]
            k2 = h * evaluate_system([func_str], {'x': x + h/2, 'y': y + k1/2})[0]
            k3 = h * evaluate_system([func_str], {'x': x + h/2, 'y': y + k2/2})[0]
            k4 = h * evaluate_system([func_str], {'x': x + h, 'y': y + k3})[0]
        except Exception as e:
            return {"success": False, "message": str(e), "steps": steps}
            
        y_next = y + (k1 + 2*k2 + 2*k3 + k4) / 6
        x_next = x + h
        
        step_data = {
            "iteration": i,
            "x": x,
            "y": y,
            "k1": k1,
            "k2": k2,
            "k3": k3,
            "k4": k4,
            "y_next": y_next
        }
        steps.append(step_data)
        
        y = y_next
        x = x_next
        
    return {
        "success": True,
        "steps": steps,
        "final_x": x,
        "final_y": y
    }

File: worker/src/methods/rk4.ts
import { evaluateSystem } from "../lib/exprEval";

export interface Rk4Step {
  iteration: number;
  x: number;
  y: number;
  k1: number;
  k2: number;
  k3: number;
  k4: number;
  y_next: number;
}

export function rungeKutta4(funcStr: string, x0: number, y0: number, h: number, stepsCount: number) {
  const steps: Rk4Step[] = [];
  let x = x0;
  let y = y0;

  for (let i = 0; i < stepsCount; i++) {
    let k1: number, k2: number, k3: number, k4: number;
    try {
      k1 = h * evaluateSystem([funcStr], { x, y })[0];
      k2 = h * evaluateSystem([funcStr], { x: x + h / 2, y: y + k1 / 2 })[0];
      k3 = h * evaluateSystem([funcStr], { x: x + h / 2, y: y + k2 / 2 })[0];
      k4 = h * evaluateSystem([funcStr], { x: x + h, y: y + k3 })[0];
    } catch (e) {
      return { success: false, message: (e as Error).message, steps };
    }

    const yNext = y + (k1 + 2 * k2 + 2 * k3 + k4) / 6;
    const xNext = x + h;

    steps.push({ iteration: i, x, y, k1, k2, k3, k4, y_next: yNext });

    y = yNext;
    x = xNext;
  }

  return { success: true, steps, final_x: x, final_y: y };
}
